src/dbnode/persist/fs/merger.go

```diff
@@ -0,0 +1,246 @@
+		encoderPool:   encoderPool,
+	}
+}
+
+func (m *Merger) Merge(
+	mergeWith MergeWith,
+	ns namespace.Metadata,
+	shard uint32,
+	blockStart xtime.UnixNano,
+	blockSize time.Duration,
+	flushPreparer persist.FlushPreparer,
+	nsCtx namespace.Context,
+) error {
+	var multiErr xerrors.MultiError
+
+	reader := m.reader
+	srPool := m.srPool
+	multiIterPool := m.multiIterPool
+	identPool := m.identPool
+	encoderPool := m.encoderPool
+
+	startTime := blockStart.ToTime()
+	openOpts := DataReaderOpenOptions{
+		Identifier: FileSetFileIdentifier{
+			Namespace:  ns.ID(),
+			Shard:      shard,
+			BlockStart: startTime,
+		},
+	}
+	if err := reader.Open(openOpts); err != nil {
+		multiErr = multiErr.Add(err)
+		return multiErr.FinalError()
+	}
+	defer reader.Close()
+
+	prepareOpts := persist.DataPrepareOptions{
+		NamespaceMetadata: ns,
+		Shard:             shard,
+		BlockStart:        startTime,
+		DeleteIfExists:    false,
+	}
+	prepared, err := flushPreparer.PrepareData(prepareOpts)
+	if err != nil {
+		multiErr = multiErr.Add(err)
+		return multiErr.FinalError()
+	}
+
+	// Writing data for a block is done in two stages. The first stage is
+	// to loop through series on disk and merge it with what's in the merge
+	// target. The second stage is to persist the rest of the series in the
+	// merge target that was not persisted in the first stage.
+
+	// First stage: loop through series on disk.
+	for id, tagsIter, data, _, err := reader.Read(); err != io.EOF; id, tagsIter, data, _, err = reader.Read() {
+		if err != nil {
+			multiErr = multiErr.Add(err)
+			return multiErr.FinalError()
+		}
+
+		// There will be at most two BlockReader slices: one for disk data
+		// and one for in memory data.
+		brs := make([][]xio.BlockReader, 0, 2)
+
+		// Create BlockReader slice out of disk data.
+		br := blockReaderFromData(data, srPool, startTime, blockSize)
+		brs = append(brs, []xio.BlockReader{br})
+
+		// Check if this series is in memory (and thus requires merging).
+		encoded, hasData, err := mergeWith.Read(id, blockStart, nsCtx)
+		if err != nil {
+			multiErr = multiErr.Add(err)
+			return multiErr.FinalError()
+		}
+
+		if hasData {
+			brs = append(brs, encoded)
+		}
+
+		mergedIter := multiIterPool.Get()
+		mergedIter.ResetSliceOfSlices(xio.NewReaderSliceOfSlicesFromBlockReadersIterator(brs), nsCtx.Schema)
+		defer mergedIter.Close()
+
+		tags, err := convert.TagsFromTagsIter(id, tagsIter, identPool)
+		tagsIter.Close()
+		if err != nil {
+			multiErr = multiErr.Add(err)
+			return multiErr.FinalError()
+		}
+		if err := persistIter(prepared.Persist, mergedIter, id, tags, encoderPool); err != nil {
+			multiErr = multiErr.Add(err)
+			return multiErr.FinalError()
+		}
+	}
+
+	// Second stage: loop through rest of the merge target that was not captured
+	// in the first stage.
+	err = mergeWith.ForEachRemaining(blockStart, func(seriesID ident.ID, tags ident.Tags) bool {
+		encoded, hasData, err := mergeWith.Read(seriesID, blockStart, nsCtx)
```

Comment:
This works although I wonder if the data should just be included and passed to the function so that you don't have to call into `mergeWith` again (which may have to look up the same data it had access to a moment again when it called this function)

Reply:
What do you mean by "the same data"? `mergeWith.Read` should only be called once per seriesID/blockStart combination for the merge target data.

Reply:
Makes sense to me now 👍